### PR TITLE
Improve windows support

### PR DESCRIPTION
**Pull request by matthiasng**
## Description
Flannel is using rakelkar/gonetsh on windows to interact with the network api.
rakelkar/gonetsh also pulls in bhendo/go-powershell as dependency. Both repositories are looking fine but not actively maintained.

This PR removes the dependencies to both repositories and implements all required api call with powershell inside flannel. Running powershell commands and parsing the output is really easy (See 7ce339f39308090ce26e11fc9b0bb54b ``func runPowerShellCommand``). Also maintaining the code shouldn't be a problem for inexperienced powershell users (Thanks to Microsoft's good documentation).

#1209 will also replace rakelkar/gonetsh but how @rajatchopra already mentions in [this comment](https://github.com/coreos/flannel/pull/1209#issuecomment-548057833) then flannel relies on another dependency.

This PR makes PR #1209 and #1180 obsolete and fixes issue #944

### go test
Run `go test ./...` on Windows

before:
```
ok      github.com/coreos/flannel/pkg/ip        0.068s  coverage: 33.8% of statements
ok      github.com/coreos/flannel/subnet        0.117s  coverage: 20.0% of statements
ok      github.com/coreos/flannel/subnet/etcdv2 0.200s  coverage: 68.2% of statements
ok      github.com/coreos/flannel/subnet/kube   0.409s  coverage: 7.3% of statements
```
after:
```
ok      github.com/coreos/flannel/pkg/ip        (cached)        coverage: 61.5% of statements
ok      github.com/coreos/flannel/pkg/routing   (cached)        coverage: 83.3% of statements
ok      github.com/coreos/flannel/subnet        (cached)        coverage: 20.0% of statements
ok      github.com/coreos/flannel/subnet/etcdv2 (cached)        coverage: 68.2% of statements
ok      github.com/coreos/flannel/subnet/kube   (cached)        coverage: 7.3% of statements
```

### Manual
Tested with [Kubernetes v1.17.0](https://kubernetes.io/docs/setup/release/notes/)

1. [install](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/) linux master node
2. [prepare](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-nodes/) windows node on windows server 2019 x64 (german version)
    - Add "-v=10" to `$flannelArgs` in [KubeClusterHelper.psm1](https://github.com/kubernetes-sigs/sig-windows-tools/blob/master/kubeadm/KubeClusterHelper.psm1#L334)
    - Join command `.\KubeCluster.ps1 -ConfigFile .\Kubeclustervxlan.json -join` will fail
        ```
        ...
        Waiting for the Network (vxlan0) to be created by flanneld
        Waiting for the Network (vxlan0) to be created by flanneld
        Waiting for the Network (vxlan0) to be created by flanneld
        Waiting for the Network (vxlan0) to be created by flanneld
        Fail to create the network[(vxlan0)] in 60 seconds
        In C:\Users\Administrator\sig-windows-tools-master\kubeadm\KubeClusterHelper.psm1:250 Zeichen:13
        +             throw "Fail to create the network[($NetworkName)] in $wai ...
        +             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            + CategoryInfo          : OperationStopped: (Fail to create ...] in 60 seconds:String) [], RuntimeException
            + FullyQualifiedErrorId : Fail to create the network[(vxlan0)] in 60 seconds
        ```
        
        Check flannel logs (Default: `C:\ProgramData\Kubernetes\logs\flanneld\flanneldsvc.log`)
        ```
        OnStart - Successfully started the service
        I0317 23:30:02.964004    1116 main.go:450] Searching for interface using 172.27.1.88
        I0317 23:30:03.073446    1116 main.go:210] Could not find valid interface matching 172.27.1.88: error looking up interface 172.27.1.88: no interfaces found in netsh output:
        ... <netsh output with german strings>
        E0317 23:30:03.073446    1116 main.go:234] Failed to find interface to use that matches the interfaces and/or regexes provided
    ```
1. Copy new `flanneld.exe` to `C:\ProgramData\Kubernetes`
1. Rerun join command
    ```
    # Get-HnsNetwork on Node

    ... 
    ActivityId             : 56D8A749-D0C4-492F-B66B-94C64F2A0759
    AdditionalParams       :
    CurrentEndpointCount   : 4
    DNSServerCompartment   : 4
    DrMacAddress           : 00-15-5D-04-9E-FC
    Extensions             : {@{Id=E7C3B2F0-F3C5-48DF-AF2B-10FED6D72E7A; IsEnabled=False; Name=Microsoft Windows-Filterplattform}, @{Id=E9B59CFA-2BE1-4B21-828F-B6FBDBDDC017; IsEnabled=True;
                             Name=Microsoft Azure VFP Switch Extension}, @{Id=EA24CD6C-D17A-4348-9190-09F0D5BE83DD; IsEnabled=True; Name=Microsoft-NDIS-Aufzeichnung}}
    Flags                  : 8
    Health                 : @{LastErrorCode=0; LastUpdateTime=132292638500939594}
    ID                     : 5E2837E0-1E0D-461E-833B-AFBD2AFB4583
    IPv6                   : False
    InterfaceConstraint    : @{InterfaceGuid=00000000-0000-0000-0000-000000000000}
    LayeredOn              : 49F1AE5E-2446-43DE-9C10-6765365A77E0
    MacPools               : {@{EndMacAddress=00-15-5D-6D-BF-FF; StartMacAddress=00-15-5D-6D-B0-00}}
    ManagementIP           : 172.27.1.88
    MaxConcurrentEndpoints : 4
    Name                   : vxlan0
    Policies               : {@{Type=HostRoute}, @{DestinationPrefix=10.244.0.0/24; DistributedRouterMacAddress=c6:71:81:86:b2:77; IsolationId=4096; ProviderAddress=172.27.10.47;
                             Type=RemoteSubnetRoute}, @{DestinationPrefix=10.244.1.0/24; DistributedRouterMacAddress=d6:d5:d1:51:45:54; IsolationId=4096; ProviderAddress=172.27.5.195;
                             Type=RemoteSubnetRoute}, @{DestinationPrefix=10.244.2.0/24; DistributedRouterMacAddress=46:36:3c:66:27:ad; IsolationId=4096; ProviderAddress=172.27.14.124;
                             Type=RemoteSubnetRoute}}
    Resources              : @{AdditionalParams=; AllocationOrder=1; Allocators=System.Object[]; Health=; ID=56D8A749-D0C4-492F-B66B-94C64F2A0759; PortOperationTime=0; State=1;
                             SwitchOperationTime=0; VfpOperationTime=0; parentId=CB8E7595-FD1B-4F68-8AC7-FACD47A44656}
    State                  : 1
    Subnets                : {@{AdditionalParams=; AddressPrefix=10.244.10.0/24; GatewayAddress=10.244.10.1; Health=; ID=354BE21E-C81C-44AA-874E-BEC50437DB4B; ObjectType=5;
                             Policies=System.Object[]; State=0}}
    TotalEndpoints         : 9
    Type                   : Overlay
    Version                : 38654705666
    ```
1. Run ` kubectl run -i --generator=run-pod/v1 --image mcr.microsoft.com/windows/servercore:ltsc2019 -- powershell`
    ```
    # ipconfig inside container
    
    Ethernet adapter vEthernet (c00bd267748c66abdc686846f16bf887fe6a3cc673c7c035a5260971e93e3f4a_vxlan0):

       Connection-specific DNS Suffix  . : default.svc.cluster.local
       Link-local IPv6 Address . . . . . : fe80::f9a8:418f:fa4:82bf%31
       IPv4 Address. . . . . . . . . . . : 10.244.10.8
       Subnet Mask . . . . . . . . . . . : 255.255.255.0
       Default Gateway . . . . . . . . . : 10.244.10.1
       
       
    # Get-HNSActivities on node   
    
    ...

    AdditionalParams    :
    AllocationOrder     : 10
    Allocators          : {@{AdapterNetCfgInstanceId={A2AFF206-B9AD-44B4-92DD-CEB6412A3E8C}; AdditionalParams=; AllocationOrder=0; CompartmendId=5; Connected=True; DevicelessNic=True;
                          EndpointNicGuid=9FF92877-04AC-4EEE-BBA7-67D822056290; EndpointPortGuid=DF43B43E-4320-4EFE-BDEA-060C9226024A; Health=; Hidden=False;
                          ID=F1B97FD9-1C9F-4A7E-9170-35F0A602B9BA; InterfaceGuid=A2AFF206-B9AD-44B4-92DD-CEB6412A3E8C; IsPolicy=False; IsolationId=4096; MTU=1450; MacAddress=0E-2A-0A-F4-0A-08;
                          ManagementPort=False; NicFriendlyName=c00bd267748c66abdc686846f16bf887fe6a3cc673c7c035a5260971e93e3f4a_vxlan0;
                          PreferredPortFriendlyName=AC1E92F9-F9C2-4C0E-9574-81772250CB96; State=3; SwitchId=9DAF13CD-5B7C-4AC5-9181-0CE706888E59; Tag=Endpoint Port; WaitForIpv6Interface=False;
                          nonPersistentPort=False}, @{AdditionalParams=; AllocationOrder=1; Health=; ID=84B8A4C3-1DD2-4DBF-A5B1-9BF7DCCDD993; InDefaultIsolation=False; IsPolicy=True;
                          IsolationID=4096; State=3; Tag=Isolation Policy}, @{AdditionalParams=; AllocationOrder=2; Health=; ID=1F435EA0-F550-4F8B-BF35-FB99E84819F9; IsPolicy=True;
                          Rules=System.Object[]; State=3; Tag=ACL Policy}, @{AdditionalParams=; AllocationOrder=3; CA=10.244.10.8; Health=; ID=D5918787-0338-4263-9DE4-B3415EE05A33; IsLocal=True;
                          IsPolicy=True; PA=172.27.1.88; State=3; Tag=VNET Policy}...}
    Health              : @{LastErrorCode=0; LastUpdateTime=132292672678199120}
    ID                  : 6D609961-68A3-4A28-AE4B-AF8EA06DFBC2
    PortOperationTime   : 0
    State               : 1
    SwitchOperationTime : 0
    VfpOperationTime    : 0
    parentId            : 56D8A749-D0C4-492F-B66B-94C64F2A0759
    ```
1. ping another container

## Todos
- [x] Tests (go tests and manual test on Windows server 2019 Standard with kubeadm)
- [x] Documentation (not necessary)
- [x] Release note (not necessary)

## Release Note

```release-note
None required
```

### CODE REVIEW

**Comment by rajatchopra** [comment]
@matthiasng Can you rebase?

**Comment by matthiasng** [comment]
@rajatchopra squash and rebase done

**Comment by rajatchopra** [comment]
@madhanrm @ksubrmnn Can we get a review from any of you? I do not have infra to test flannel on windows.

**Comment by ksubrmnn** [comment]
@JocelynBerrendonner

**Comment by JocelynBerrendonner** [comment]
@dineshgovindasamy FYI.

**Comment by JocelynBerrendonner** [comment]
@matthiasng what did the test look like on your change? (I am asking because they are marked as pending in the description)

**Comment by matthiasng** [comment]
@JocelynBerrendonner I added my test cases to the description.

I'm new to k8s CNI so im not sure if they are enought.

File: pkg/ip/ipnet.go [hunk 1]
```diff
@@ -31,7 +31,13 @@ func FromBytes(ip []byte) IP4 {
 }
 
 func FromIP(ip net.IP) IP4 {
-	return FromBytes(ip.To4())
+	ipv4 := ip.To4()
+
+	if ipv4 == nil {
+		panic("Address is no IPv4 address")
```
Comment (elweb9858) [inline pkg/ip/ipnet.go:37]: nit: spelling *not

File: pkg/routing/router_windows_test.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		t.Fatal(err)
+	}
+	if len(allRoutes) == 0 {
+		t.Fatal("len(routes) == 0")
+	}
+
+	destinationSubnet := allRoutes[0].DestinationSubnet
+	gatewayAddress := net.ParseIP("192.168.199.123")
+
+	//
```
Comment (elweb9858) [inline pkg/routing/router_windows_test.go:55]: remove empty comment

File: pkg/routing/router_windows_test.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	firstRouteLen := len(routes)
+	if len(routes) == 0 {
+		t.Fatal("len(routes) == 0")
+	}
+
+	//
```
Comment (elweb9858) [inline pkg/routing/router_windows_test.go:66]: remove empty comment

File: pkg/routing/router_windows_test.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		t.Fatal(err)
+	}
+
+	expectedLen := firstRouteLen + 1
+	givenLen := len(routes)
+	if givenLen != expectedLen {
+		t.Fatalf("givenLen:%d != expectedLen:%d", givenLen, expectedLen)
+	}
+
+	//
```
Comment (elweb9858) [inline pkg/routing/router_windows_test.go:83]: same as above; remove empty comment

**Comment by JocelynBerrendonner** [comment]
LGTM

@rajatchopra : @elweb9858 handled Windows testing and signoff from my team's perspective. So, this should be good to go from our perspective!

Thanks

**Comment by matthiasng** [comment]
Typo fixed and empty comments are gone.

**Comment by elweb9858** [comment]
Manual tests passed on windows:
19h119h1
RS5RS5

**Comment by matthiasng** [comment]
@elweb9858 can you tell me where i can find these test ?

**Comment by luthermonson** [comment]
👍  on this, just forked master and cherry-picked this hash and it worked perfectly on a windows cluster. LGTM

**Pull request merged**